src/Storages/StorageMemory.cpp

```diff
@@ -114,8 +108,9 @@ class MemoryBlockOutputStream : public IBlockOutputStream
     void write(const Block & block) override
     {
         metadata_snapshot->check(block, true);
-        std::lock_guard lock(storage.mutex);
-        storage.data.push_back(block);
+        auto new_data = std::make_unique<BlocksList>(*(storage.data.get()));
+        new_data->push_back(block);
+        storage.data.set(std::move(new_data));
```

Comment:
There is going to be a problem due to a race condition here. Because we do not employ any CAS technique here, data can be lost when concurrent INSERTs each add a block to its own copy of data.

Reply:
So, in order to solve this problem, we still need a table-level lock?

Reply:
So, in order to solve this problem, we still need a table-level lock?